### PR TITLE
236 - SIM-Add signatures to transactions 

### PR DESCRIPTION
**Pull request by bradleySuira**
Fixes: 
- #236 
- #369 
- #353 
- #246 
- #371 
- #286 

# What
- Added transaction encoding, signing utils, and tests to the frontend.
- Added transaction signature to the frontend when calling a contract deployment and when calling a contract method
- Added new RPC endpoint `send_raw_transaction` that verifies the transaction signature and decodes the data 
- Refactored the integration tests to use the `send_raw_transaction` as well.

# Why

Our RPC server should be ethereum-compatible. One key feature towards that is to send transactions in the same way any account does it in ethereum.

The 

# Testing done

- run unit testing in frontend
- verify transaction signatures in the front end and backend

# Decisions made

- We keep a single lib in frontend. In this case, Viem does not add another
- I avoided to refactor the whole code fully, more improvements should be applied to the frontend to improve the way we are structuring the code.

# Checks

- [x] I have tested this code
- [x] I have reviewed my own PR
- [x] I have created an issue for this PR
- [x] I have set the PR name to the issue name

# User facing release notes
- The RPC server is not ethereum-compatible to send transactions and to deploy contracts. The method send_raw_transaction has been implemented to send serialized signed transactions. The previous way of sending transactions and deploying contracts has been deprecated and those methods are going to be removed in the next versions.

# Other notes
- More unit tests should be added both in the backend and in the frontend

### CODE REVIEW

**Review by cristiam86 — CHANGES_REQUESTED** [review]
Good job overall :)

File: backend/errors/errors.py [hunk 1]
```diff
@@ -89,3 +89,19 @@ def __init__(
         self.id = id
         self.message = message
         super().__init__(self.message)
+
+
+class InvalidTransactionError(Exception):
+    """Exception raised when the given transaction is not valid or can be validated ."""
```
[the remark this answers is not in the document]
Comment (bradleySuira) [reply backend/errors/errors.py:95]: fixed

File: backend/protocol_rpc/endpoints.py [hunk 2]
```diff
@@ -311,6 +313,61 @@ def clear_db_tables(db_client: DBClient, tables: list) -> dict:
...
+
+    from_address = decoded_transaction["from"]
+    if not address_is_in_correct_format(from_address):
+        raise InvalidAddressError(from_address)
+
+    to_address = decoded_transaction["to"]
+    
+    transaction_data = {}
+    result = {}
+    if to_address and to_address != "0x":
```
[the remark this answers is not in the document]
Comment (bradleySuira) [reply backend/protocol_rpc/endpoints.py:335]: in this case the to address is the contract that we want to call, if we want to deploy a contract there is no `to` address, so we can infer that is a deployment, that's the reason for it, in the same method we have like the two others `deploy_intelligent_contract` and `call_contract_function`. if it's like you say if we have a to address we can call to `call_contract_function ` if not call `deploy_intelligent_contract`

File: frontend/test/unit/services/signTransaction.test.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import {
+  signTransaction,
+  getPrivateKey,
+  getAccountFromPrivatekey,
+  getSignedTransactionAddress
+} from '../../../src/utils'
+import { describe, expect, it, vi, afterEach, beforeEach } from 'vitest'
+
+describe('JsonRprService', () => {
```
[the remark this answers is not in the document]
Comment (bradleySuira) [reply frontend/test/unit/services/signTransaction.test.ts:9]: fixed

File: backend/protocol_rpc/endpoints.py [hunk 4]
```diff
@@ -311,6 +313,61 @@ def clear_db_tables(db_client: DBClient, tables: list) -> dict:
...
+        }
+        result["contract_address"] = to_address
+
+    transaction_id = transactions_processor.insert_transaction(
+        from_address, to_address, transaction_data, 0, 2
+    )
+
+    print("decoded_transaction", decoded_transaction, transaction_data)
+
+    transaction_id = transactions_processor.insert_transaction(
```
[the remark this answers is not in the document]
Comment (bradleySuira) [reply backend/protocol_rpc/endpoints.py:363]: fixed should be only one insert

**Comment by cristiam86** [comment]
@bradleySuira let's do the following:
- Remove the unused code from the front-end
- Encode and sign the transactions just like Ethereum does it
- Cristiam will fix and finish the backend side of it

**Comment by bradleySuira** [comment]
> @bradleySuira let's do the following:
> 
> * Remove the unused code from the front-end
> * Encode and sign the transactions just like Ethereum does it
> * Cristiam will fix and finish the backend side of it

done @cristiam86 you can pull it

File: backend/protocol_rpc/transaction_utils.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
+# rpc/transaction_utils.py
+
+from typing import Any
+from eth_account._utils.typed_transactions import TypedTransaction
```
Comment (AgustinRamiroDiaz) [inline backend/protocol_rpc/transaction_utils.py:4]:
```suggestion
from eth_account.typed_transactions import TypedTransaction
```

File: backend/database_handler/accounts_manager.py [hunk 6]
```diff
@@ -23,6 +25,14 @@ def _parse_account_data(self, account_data: dict) -> dict:
             "updated_at": account_data["updated_at"].isoformat(),
         }
 
+    def create_new_account(self, balance: int = 0) -> Account:
```
Comment (cristiam86) [inline backend/database_handler/accounts_manager.py:28]: This backend account creation is used when generating the intelligent contract account.

File: backend/node/genvm/base.py [hunk 7]
```diff
@@ -204,14 +203,67 @@ def get_contract_schema(contract_code: str) -> dict:
...
                 "class": class_name,
                 "methods": methods,
                 "variables": variables,
+                "abi": abi,
             }
 
         return contract_schema
 
+    @staticmethod
+    def get_abi_param_type(param_type: str) -> str:
```
Comment (cristiam86) [inline backend/node/genvm/base.py:218]: Now we also generate an Ethereum-compatible ABI for the intelligent contract interface.
The current way of doing it (https://github.com/yeagerai/genlayer-simulator/pull/315/files#diff-f9c09ce67d4e8dddb05fbad6e64f19e0462222f8ab24310cc8edaa016f32e7a5R157) will be deprecated when the frontend gets refactored.

File: backend/protocol_rpc/endpoints.py [hunk 8]
```diff
@@ -312,6 +271,70 @@ def clear_db_tables(db_client: DBClient, tables: list) -> dict:
     db_client.clear_tables(tables)
 
 
+def send_raw_transaction(
```
Comment (cristiam86) [inline backend/protocol_rpc/endpoints.py:274]: This is the new method to deploy contracts and call their functions

File: backend/protocol_rpc/endpoints.py [hunk 9]
```diff
@@ -312,6 +271,70 @@ def clear_db_tables(db_client: DBClient, tables: list) -> dict:
...
+    )
+    if not transaction_signature_valid:
+        raise InvalidTransactionError("Transaction signature verification failed")
+
+    to_address = decoded_transaction["to"]
+
+    transaction_data = {}
+    result = {}
+    transaction_type = -1
+    if to_address and to_address != "0x":
```
Comment (cristiam86) [inline backend/protocol_rpc/endpoints.py:303]: Here we decide if the transaction is a contract deployment or a contract method calling.
The node should do this. Here, we should store the call data encoded in the transactions database.

File: backend/protocol_rpc/transactions_parser.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    return res
+
+
+def transaction_has_valid_signature(raw_tx: str, decoded_tx: dict) -> bool:
+    recovered_address = Account.recover_transaction(raw_tx)
+    # Compare the recovered address with the 'from' address in the transaction
+    return recovered_address == decoded_tx["from"]
+
+
+def decode_data_field(input: HexBytes) -> str:
```
Comment (cristiam86) [inline backend/protocol_rpc/transactions_parser.py:46]: This is our custom way of decoding the data based on what we did in the frontend here: https://github.com/yeagerai/genlayer-simulator/pull/315/files#diff-86f9fcc5ab8dbb626a7814b69c5a32b5290572002ccc0405dcb608465faca384R185 
and here: https://github.com/yeagerai/genlayer-simulator/pull/315/files#diff-86f9fcc5ab8dbb626a7814b69c5a32b5290572002ccc0405dcb608465faca384R283

File: frontend/src/stores/contracts.ts [hunk 11]
```diff
@@ -177,12 +177,25 @@ export const useContractsStore = defineStore('contractsStore', () => {
...
-        contractAddress: contract?.address || '',
-        method,
-        params,
-      });
+
+      if (!accountsStore.currentPrivateKey) {
+        return false;
+      }
+
+      const methodParamsAsString = JSON.stringify(params)
```
Comment (cristiam86) [inline frontend/src/stores/contracts.ts:185]: This is how we serialize the call contract transaction in the frontend

File: frontend/src/stores/contracts.ts [hunk 12]
```diff
@@ -177,12 +177,25 @@ export const useContractsStore = defineStore('contractsStore', () => {
...
+      const data = toRlp([
+        toHex(method),
+        toHex(methodParamsAsString)
+      ])
+      
+      const to = contract?.address as Address ?? null;
+      const signed = await signTransaction(
```
Comment (cristiam86) [inline frontend/src/stores/contracts.ts:192]: This uses the internal Viem method to sign the transaction. Here is this function implementation: https://github.com/yeagerai/genlayer-simulator/pull/315/files#diff-fa9662a1666b1c93a605e8a50d112a7884f29d058380bff14cd13366fa9b9cc4R5

File: frontend/src/stores/contracts.ts [hunk 13]
```diff
@@ -260,41 +273,50 @@ export const useContractsStore = defineStore('contractsStore', () => {
...
-            } else {
-              throw new Error(
-                typeof result?.message === 'string'
-                  ? result.message
-                  : 'Error Deploying the contract',
-              );
+            if (!accountsStore.currentPrivateKey) {
+              return
             }
+              const constructorParamsAsString = JSON.stringify(constructorParams)
```
Comment (cristiam86) [inline frontend/src/stores/contracts.ts:283]: Serialization of the deploy contract data

File: frontend/src/utils/transactions.ts [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import { getAccountFromPrivatekey } from './accounts';
+
+import type {Address} from '@/types';
+
+export async function signTransaction(
```
Comment (cristiam86) [inline frontend/src/utils/transactions.ts:5]: Using Viem to sign the transaction with the data custom serialized

File: tests/common/request.py [hunk 15]
```diff
@@ -41,10 +43,29 @@ def get_transaction_by_id(transaction_id: str):
     return raw_response.json()
 
 
-def post_request_and_wait_for_finalization(
-    payload: dict, interval: int = 10, retries: int = 15
+def call_contract_method(
```
Comment (cristiam86) [inline tests/common/request.py:46]: New backend integration tests helpers to use the send_raw_transaction RPC method

File: tests/common/transactions.py [hunk 16]
```diff
@@ -0,0 +1,24 @@
+from eth_account import Account
+from eth_utils import to_hex
+import rlp
+from eth_account._utils.legacy_transactions import Transaction
+
+
+def encode_transaction_data(data: list) -> str:
```
Comment (cristiam86) [inline tests/common/transactions.py:7]: This is the custom way of encoding the transactions in the backend integration tests that produces the same encoding than the frontend code

**Review by cristiam86 — COMMENTED** [review]
@MuncleUscles I have added some notes to help you review this implementation

**Review by cristiam86 — APPROVED** [review]

File: backend/node/base.py [hunk 17]
```diff
@@ -48,39 +47,59 @@ async def exec_transaction(self, transaction: dict):
...
+            return self.parse_transaction_execution_receipt_on_error(e)
+
+        return self.parse_transaction_execution_receipt_on_success(receipt)
 
     async def run_contract(
         self, from_address: str, function_name: str, args: list
     ) -> dict:
         receipt = None
         try:
+            parsed_args = json.loads(args)
```
Comment (AgustinRamiroDiaz) [inline backend/node/base.py:93]: was this a bug that we didn't know about?

File: backend/protocol_rpc/transactions_parser.py [hunk 18]
```diff
@@ -0,0 +1,78 @@
+# rpc/transaction_utils.py
+
+from typing import Any
```
Comment (AgustinRamiroDiaz) [inline backend/protocol_rpc/transactions_parser.py:3]:
```suggestion
```

File: backend/protocol_rpc/transactions_parser.py [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    data_bytes = HexBytes(data)
+    data_decoded = rlp.decode(data_bytes, MethodCallTransactionPayload)
+
+    return {
+        "function_name": decode_data_field(data_decoded["function_name"]),
+        "function_args": decode_data_field(data_decoded["function_args"]),
+    }
+
+
+def decode_deployment_data(data: str) -> str:
```
Comment (AgustinRamiroDiaz) [inline backend/protocol_rpc/transactions_parser.py:61]:
```suggestion
def decode_deployment_data(data: str) -> dict:
```

File: backend/protocol_rpc/transactions_parser.py [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    # Compare the recovered address with the 'from' address in the transaction
+    return recovered_address == decoded_tx["from"]
+
+
+def decode_data_field(input: HexBytes) -> str:
+    param = input.hex().replace("0x", "")
+    return codecs.decode(param, "hex").decode("utf-8")
+
+
+def decode_method_call_data(data: str) -> str:
```
Comment (AgustinRamiroDiaz) [inline backend/protocol_rpc/transactions_parser.py:51]:
```suggestion
def decode_method_call_data(data: str) -> dict:
```

File: backend/protocol_rpc/transactions_parser.py (hunk 10 above)
Comment (AgustinRamiroDiaz) [reply backend/protocol_rpc/transactions_parser.py:46]: @cristiam86 Is this documented somewhere? seems like something we might document about "how transactions are signed"

File: backend/protocol_rpc/transactions_parser.py [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+import codecs
+from eth_account._utils.legacy_transactions import Transaction, vrs_from
+from eth_account._utils.signing import hash_of_signed_transaction
+from eth_account import Account
+from eth_utils import to_checksum_address
+from hexbytes import HexBytes
+from rlp.sedes import binary
+
+
+def decode_signed_transaction(raw_tx):
```
Comment (AgustinRamiroDiaz) [inline backend/protocol_rpc/transactions_parser.py:14]: Is there a way to make this more readable? I personally find it hard to read all the `tx`, `txn` and `res`.

**Review by AgustinRamiroDiaz — APPROVED** [review]
Great work! Everything looks good, I've left some doubts and nits

**Review by AgustinRamiroDiaz — APPROVED** [review]
Great work! I've reviewed the backend and I've left some doubts and nits

File: backend/database_handler/accounts_manager.py (hunk 6 above)
Comment (cristiam86) [reply backend/database_handler/accounts_manager.py:28]: I have checked this and it is how it works on Ethereum.

File: backend/node/genvm/base.py (hunk 7 above)
Comment (cristiam86) [reply backend/node/genvm/base.py:218]: Let's address the ABI creation later when the GenVM and the Node make some progress.

File: backend/protocol_rpc/transactions_parser.py (hunk 10 above)
Comment (cristiam86) [reply backend/protocol_rpc/transactions_parser.py:46]: https://github.com/orgs/yeagerai/projects/2/views/2?filterQuery=-status%3Adone+docs&pane=issue&itemId=74589262

File: backend/node/genvm/base.py [hunk 22]
```diff
@@ -204,14 +203,67 @@ def get_contract_schema(contract_code: str) -> dict:
...
+        if param_type == "dict":
+            return "bytes"
+        if param_type == "list":
+            return "bytes"
+        if param_type == "None":
+            return "None"
+        return param_type
+
+    @staticmethod
+    def generate_abi_from_schema_methods(contract_schema_methods: dict) -> list:
```
[the remark this answers is not in the document]
Comment (cristiam86) [reply backend/node/genvm/base.py:234]: True, I'll start tidying when developing next issues

File: backend/node/genvm/base.py [hunk 23]
```diff
@@ -204,14 +203,67 @@ def get_contract_schema(contract_code: str) -> dict:
...
+        abi = []
+
+        for method_name, method_info in contract_schema_methods.items():
+            abi_entry = {
+                "name": method_name,
+                "type": "function",
+                "inputs": [],
+                "outputs": [],
+            }
```
[the remark this answers is not in the document]
Comment (cristiam86) [reply backend/node/genvm/base.py:243]: Let's please address abi generation layer

File: backend/node/genvm/base.py [hunk 24]
```diff
@@ -204,14 +203,67 @@ def get_contract_schema(contract_code: str) -> dict:
...
+
+            for input_name, input_type in method_info["inputs"].items():
+                abi_entry["inputs"].append(
+                    {"name": input_name, "type": GenVM.get_abi_param_type(input_type)}
+                )
+
+            if method_info["output"]:
```
[the remark this answers is not in the document]
Comment (cristiam86) [reply backend/node/genvm/base.py:250]: https://github.com/orgs/yeagerai/projects/2/views/2?filterQuery=-status%3Adone&pane=issue&itemId=74589630

File: backend/protocol_rpc/transactions_parser.py (hunk 21 above)
Comment (cristiam86) [reply backend/protocol_rpc/transactions_parser.py:14]: done!

File: backend/node/genvm/base.py [hunk 25]
```diff
@@ -64,43 +66,57 @@ def _get_contract_class_name(contract_code: str) -> str:
...
+            class_name,
+            method_name,
+            args,
+            self.contract_runner.gas_used,
+            self.contract_runner.mode,
+            encoded_object,
+            self.contract_runner.node_config,
+            self.contract_runner.eq_outputs,
+            execution_result,
+            error,
```
Comment (AgustinRamiroDiaz) [inline backend/node/genvm/base.py:88]: nit: I'd set them through their keywords in order to not depend on the order of the class fields

```suggestion
        return Receipt(
            class_name=class_name,
            method=method_name,
            args=args,
            gas_used=self.contract_runner.gas_used,
            mode=self.contract_runner.mode,
            contract_state=encoded_object,
            node_config=self.contract_runner.node_config,
            eq_outputs=self.contract_runner.eq_outputs,
            execution_result=execution_result,
            error=error,
```

File: backend/node/base.py [hunk 26]
```diff
@@ -16,7 +17,7 @@ def __init__(
         provider: str,
         model: str,
         config: dict,
-        leader_receipt: Optional[dict] = None,
+        leader_receipt: Optional[Receipt] = None,
     ):
         self.validator_mode = validator_mode
```
Comment (AgustinRamiroDiaz) [inline backend/node/base.py:22]: on some parts of the code `validator_mode` is a string, and in others it's an `ExecutionMode` enum. I think they should all be `ExecutionMode`. 

Be careful on places where the serialization to `str` is required (like when returning a JSON). Remember to cast the enum to its `.value` on these "endpoints"

File: backend/node/base.py [hunk 27]
```diff
@@ -4,6 +4,7 @@
 
 from backend.node.genvm.base import GenVM
 from backend.database_handler.contract_snapshot import ContractSnapshot
+from backend.node.genvm.types import Receipt, ExecutionResultStatus, ExecutionMode, Vote
```
Comment (AgustinRamiroDiaz) [inline backend/node/base.py:7]:
```suggestion
from backend.node.genvm.types import Receipt, ExecutionMode, Vote
```

**Review by AgustinRamiroDiaz — APPROVED** [review]
I think the typing changes look :100: . They are also great to help us understand dependencies among different packages

Remember to review the merge conflicts

**Review by MuncleUscles — APPROVED** [review]

File: backend/node/genvm/base.py (hunk 25 above)
Comment (cristiam86) [reply backend/node/genvm/base.py]: Done!

File: backend/node/base.py (hunk 26 above)
Comment (cristiam86) [reply backend/node/base.py:22]: Done!

File: backend/node/base.py (hunk 27 above)
Comment (cristiam86) [reply backend/node/base.py]: Done!

**Review by AgustinRamiroDiaz — APPROVED** [review]

**Review by AgustinRamiroDiaz — APPROVED** [review]

**Pull request merged**